'''Name of file is to convert BAM to BED,but we are assuming we start with a SAM file(and add in the option of converting a BAM file)SAM file formatA	QNAME	String	Query template NAMEB	FLAG	Int	bitwise FLAGC	RNAME	String	References sequence NAMED	POS	Int	1- based leftmost mapping POSitionE	MAPQ	Int	MAPping QualityF	CIGAR	String	CIGAR StringG	RNEXT	String	Ref. name of the mate/next readH	PNEXT	Int	Position of the mate/next readI	TLEN	Int	observed Template LENgthJ	SEQ	String	segment SEQuenceK	QUAL	String	ASCII of Phred-scaled base QUALity+33BED file format1) chromosome2) chrom start (0 based))3) chrom end4) Name 5) Score6) Strand7) thick start8) thick end9) item rgb10) blockcount (number of exons))11) blocksizes12) block startsWant to retrieve the following characteristics fromSAM file and write to a BED file in the following format, USING PYSAM:1: 2: D-13: (D-1) + I4: C5: "."6:  '''import pysambamfilepath=bedfilepath= import pysamsamfilepath = ""samfile = pysam.AligmentFile(samfilepath,"r")samfile_iter = samfile.fetch()for alignment in samfile_iter:    if alignment.is_paired:        print(f"{alignment.reference_name}\t{alignment.query_alignment_start}\t{alignment.query_alignment_end}\t{alignment.query_name}\t{alignment.mapping_quality}\t.\n")        print(alignment)        